npeps/npep-122.md

```diff
@@ -158,7 +158,344 @@ Fourth, the ANP subject allows using pod selectors, while tenancy use cases only
+- Users that don't care about tenancy can just ignore this CRD
+- We can throw it away if we want to change API again
+
+
+#### Final suggestion
+
+Considering we agree with option 4 from the previous section on priorities specification, we can outline further details here.
+
+For
+```yaml
+kind: NetworkTenancy
+spec:
+  tenancyLabels: ["user"]
+  precedence: ANP/BANP
+  action: Allow/Skip
+```
+there are only 3 useful combinations of precedence+action (BANP+Skip is the same as BANP+Allow), so we can express these
+options in a different way, e.g. define 3 modes
+- always-allow (ANP+Allow)
+- delegate-to-NP (ANP+Skip)
+- baseline-allow (BANP+Allow)
+
+<details>
+<summary>Full yaml examples (with the initial fields, will be updates as we agree on the final CRD format)</summary>
+
+* 4.1 "overridable isolation"
+```yaml
+kind: NetworkTenancy
+spec:
+  tenancyLabels:
+    - "user"
+  precedence: BANP
+  action: Allow
+---
+kind: BaselineAdminNetworkPolicy
+spec:
+  subject:
+    namespaces:
+      matchExpression:
+        - key: user
+          operator: Exists
+  ingress:
+    - action: Deny
+      from:
+        - namespaces: {}
+  egress:
+    - action: Deny
+      to:
+        - namespaces: {}
+```
+BANP can also be replaced with deny-all BANP
+```yaml
+kind: BaselineAdminNetworkPolicy
+spec:
+  subject:
+    namespaces: {}
+  ingress:
+    - action: Deny
+      from:
+        - namespaces: {}
+  egress:
+    - action: Deny
+      to:
+        - namespaces: {}
+```
+
+* 4.2 strict isolation
+```yaml
+kind: NetworkTenancy
+spec:
+  tenancyLabels:
+    - "user"
+  precedence: ANP
+  action: Skip
+---
+kind: AdminNetworkPolicy
+spec:
+  priority: 1
+  subject:
+    namespaces:
+      matchExpression:
+        - key: user
+          operator: Exists
+  ingress:
+    - action: Deny
+      from:
+        - namespaces: {}
+  egress:
+    - action: Deny
+      to:
+        - namespaces: {}
+```
+
+* 4.3 Allow internal connections for tenants
+```yaml
+kind: NetworkTenancy
+spec:
+  tenancyLabels:
+    - "user"
+  precedence: ANP
+  action: Allow
+---
+kind: BaselineAdminNetworkPolicy
+spec:
+  subject:
+    namespaces:
+      matchExpression:
+        - key: user
+          operator: Exists
+  ingress:
+    - action: Deny
+      from:
+        - namespaces: {}
+  egress:
+    - action: Deny
+      to:
+        - namespaces: {}
+```
+
+* 4.4 Tenants interaction with (B)ANP
+* 4.4.1 allow from monitoring + deny from not same tenant
+```yaml
+kind: NetworkTenancy
+spec:
+  tenancyLabels:
+    - "user"
+  precedence: ANP
+  action: Skip
+---
+kind: AdminNetworkPolicy
+spec:
+  priority: 1
+  subject:
+    namespaces:
+      matchExpression:
+        - key: user
+          operator: Exists
+  ingress:
+    - action: Allow
+      from:
+        - namespaces:
+            namespaceSelector:
+              matchLabels:
+                kubernetes.io/metadata.name: monitoring-ns
+    - action: Deny
+      from:
+        - namespaces: {}
+```
```

Comment:
Looking at this again:
- If the precedence is "ANP, Tenancy, NP, BANP" then the tenancy rule is ignored, because the ANP's Deny would run first
- If the precedence is "Tenancy, ANP, NP, BANP" then the Allow-from-monitoring rule is ignored, because the Tenancy's Skip rule would bypass the ANP.

Reply:
In this example it works as your second option, but Skip action is only applied to sameTenant namespaces, while monitoring is **not** sameTenant. So traffic from monitoring ns won't match tenancy rules (unless there is a tenant of which monitoring is a part, then monitoring will be passed as sameTenant)